File: lib/wip/variable_interpolation.rb
# frozen_string_literal: true

module Wip
  # Interpolates ${VAR} references the way `docker compose` does when reading
  # compose.yml: ${VAR}, ${VAR:-default}/${VAR-default}, bare $VAR, and $$ as an
  # escaped literal dollar sign. ${VAR:?err}/${VAR:+alt} aren't recognized and
  # pass through untouched, same as an unset $VAR with no default.
  module VariableInterpolation
    PATTERN = /\$\$|\$\{([A-Za-z_][A-Za-z0-9_]*)((:-|-)([^}]*))?\}|\$([A-Za-z_][A-Za-z0-9_]*)/

    def self.call(text, env)
      text.gsub(PATTERN) do |matched|
        next '$' if matched == '$$'

        name, _, operator, default, bare_name = Regexp.last_match.captures
        resolve(env[name || bare_name], operator, default)
      end
    end

    # Walks an already-parsed YAML structure and interpolates string values only —
    # real Compose interpolates YAML values, never mapping keys (its docs call this
    # out explicitly), and doing this after parsing means a substituted value can't
    # introduce YAML syntax (e.g. a literal "#" turning into a comment marker).
    def self.tree(value, env)
      case value
      when String then call(value, env)
      when Hash then value.transform_values { |v| tree(v, env) }
      when Array then value.map { |v| tree(v, env) }
      else value
      end
    end

    def self.resolve(value, operator, default)
      case operator
      when ':-' then value.nil? || value.empty? ? default.to_s : value
      when '-' then value.nil? ? default.to_s : value
      else value.to_s
      end
    end
    private_class_method :resolve
  end
end

File: src/Wip.Core/Compose/VariableInterpolation.cs
using System.Text.RegularExpressions;

namespace Wip.Compose;

/// <summary>
/// Interpolates <c>${VAR}</c> references the way <c>docker compose</c> does when reading
/// compose.yml: <c>${VAR}</c>, <c>${VAR:-default}</c> / <c>${VAR-default}</c>, bare
/// <c>$VAR</c>, and <c>$$</c> as an escaped literal dollar sign.
/// </summary>
/// <remarks>
/// <c>${VAR:?err}</c> and <c>${VAR:+alt}</c> are not recognised by the pattern at all, so —
/// unlike an unset <c>$VAR</c>, which resolves to an empty string — they pass through
/// completely unchanged, braces and all.
/// </remarks>
public static partial class VariableInterpolation
{
    public static string Call(string text, IReadOnlyDictionary<string, string> environment) =>
        Pattern().Replace(text, match => Substitute(match, environment));

    private static string Substitute(Match match, IReadOnlyDictionary<string, string> environment)
    {
        if (match.Value == "$$")
        {
            return "$";
        }

        var braced = match.Groups[1];
        var name = braced.Success ? braced.Value : match.Groups[5].Value;
        var found = environment.TryGetValue(name, out var value);

        var operatorGroup = match.Groups[3];
        if (!operatorGroup.Success)
        {
            return found ? value! : string.Empty;
        }

        var fallback = match.Groups[4].Value;
        return operatorGroup.Value switch
        {
            // ':-' also falls back for a set-but-empty value; '-' only for a missing one.
            ":-" => !found || value!.Length == 0 ? fallback : value!,
            _ => found ? value! : fallback,
        };
    }

    [GeneratedRegex(@"\$\$|\$\{([A-Za-z_][A-Za-z0-9_]*)((:-|-)([^}]*))?\}|\$([A-Za-z_][A-Za-z0-9_]*)")]
    private static partial Regex Pattern();
}
